docs/rfcs/027-pageserver-disaster-recovery.md

```diff
@@ -0,0 +1,157 @@
+process. It takes the WAL of the specified timeline, and writes its results
+into a *different* timeline, the target timeline. This two timeline scheme
+ensures that we can stil serve requests and preserves immutability guarantees
+assumed by the [generation numbers RFC], as in corruption situations we might
+indeed end up with different content than the original timeline we are storing
+data from.
+
+Where the parameters are:
+
+* `:tenant_id`: the id of the tenant to recover
+* `:timeline_id`: the id of the timeline to recover
+* `from_lsn`: The lsn at the start of our recovery process.
+  Optional, if not specified the recovery process is fully WAL based.
+  If specified, the recovery process starts from data at the *main* timeline
+  at the specific lsn (maybe we still trust that lsn). In any case, the target
+  timeline must be less advanced than `from_lsn`.
+* `to_lsn`: the end of the lsn range we want to recover. As usual, it's an
+  exclusive limit. This allows playing back the WAL to a specific moment in the
+  past, if this is wanted/needed. E.g. there might be a data-corrupting
+  pageserver crash at lsn 1234 and we don't want that so we play back until
+  1233.
+* `target`: the timeline id we are writing into. The parameter is optional;
+  if no target timeline is provided, a new timeline is created.
+  One specifies either both `from_lsn` and `target`, or neither.
+  Writing into the same timeline is forbidden.
+
+[generation numbers RFC]: https://github.com/neondatabase/neon/pull/4919
+
+### Higher level features
+
+The API endpoint described above allows for a large set of higher level
+features to be implemented.
+
+* recovery not just one timeline, but of multiple timelines. Oncluding, if
+  wanted, all timelines of the tenant.
+* recovery just until a specific moment in time, even if pageserver history is
+  completely turned off.
+* a repeated job to *detect* errors/corruption, both to verify that the backup
+  works and to verify that the backed up original is not corrupted.
+
+## Non Goals
+
+At the danger of being repetitive, the main goal of this feature is to be a
+backup method, so reliability is very important. This implies that other
+aspects like performance or space reduction are less important.
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+Most changes would happen to the pageservers.
+For the higher level features, maybe other components like the console would
+be involved.
+
+We need to make sure that the shadow timelines are not subject to the usual
+limits and billing we apply to existing timelines.
+
+## Proposed implementation
+
+The first problem to keep in mind is the reproducability of `initdb`.
+So an initial step would be to upload `initdb` snapshots to S3,
+maybe deduplicating them by hash, for a given storage/postgres version
+combination.
+
+After that, we'd have the endpoint spawn a background process which
+performs the replay of the WAL to that new timeline.
+
```

Comment:
After recovery, if we want to use the recovered timeline from an endpoint for writes as well as reads, how will that work?

Thinking of when we have original timeline A, and we run recovery timeline B, targeting the "WAL A".  If we then attach an endpoint to timeline B, and it wants to do some writes, does timeline B then need to consume from a new "WAL B"?  Perhaps the idea is that after recovery the B timeline will automatically switch from consuming WAL A to WAL B, but it would be good to make that part explicit.

Reply:
My comment https://github.com/neondatabase/neon/pull/5248#discussion_r1321851512 is along the same lines. Will make it explicit in the RFC.